enhancements/authentication/direct-external-oidc-provider.md

```diff
@@ -212,6 +242,76 @@ These changes have already been implemented, and the initial PR for them can be
 
+#### authorization.openshift.io/RestrictSubjectBindings admission plugin considerations
+
+OpenShift's default admission plugin `authorization.openshift.io/RestrictSubjectBindings`
+utilizes the OAuth stack, specifically the `User` and `Group` APIs, to ensure `RoleBinding`
+subjects do not violate any restrictions in the namespace they are going to be created in.
+These restrictions are specified by `RoleBindingRestriction` custom resources.
+
+Removing the OAuth stack entirely during external OIDC setup may result in the `authorization.openshift.io/RestrictSubjectBindings` admission plugin not functioning as expected.
+
+Anticipated impacts are:
+
+- Logs in the KAS reflecting that informers for OAuth APIs could not be started
+- Logs in the KAS reflecting that the oauth-apiserver is unreachable
+- Errors, that are not actionable from the perspective of users, occurring during validation of `RoleBinding` resources when a `RoleBindingRestriction` resource in the same namespace specifies user/group restrictions
+
+These impacts may cause concern for cluster administrators that their cluster is in an unhealthy state and may not be easily correlated to the enablement of the OIDC functionality.
+
+To ensure a holistic change in cluster behavior when enabling the OIDC functionality, and reduce potentially concerning behavior in the control plane components, it is proposed that changes are made to:
+
+- Disable the `authorization.openshift.io/RestrictSubjectBindings` and `authorization.openshift.io/ValidateRoleBindingRestriction` admission plugins
+- Remove the `rolebindingrestrictions.authorization.openshift.io` CustomResourceDefinition
+- Block OIDC enablement while any `RoleBindingRestriction` resources exist; This will be communicated in the `Authentication` resource via the `OIDCConfig` status field
+
+##### Changes to the kube-apiserver
+
+The OpenShift-specific patch to the kube-apiserver that adds this admission plugin is found here:
+https://github.com/openshift/kubernetes/blob/3c62f738ce74a624d46b4f73f25d6c15b3a80a2b/openshift-kube-apiserver/openshiftkubeapiserver/patch.go#L71
+
+In order to prevent misleading logs about informers that failed to start or failure to connect to the oauth-apiserver, the following changes to this patch are to be made:
+
+- Informers for the `Group` API are only configured and started as part of the first run of the `authorization.openshift.io/RestrictSubjectBindings` admission plugin validation loop. This makes it such that the informer will not be configured or attempt to start when the admission plugin is disabled.
+- The post-start hook that checks for oauth-apiserver connectivity will be skipped if the `Authentication` resource `.spec.type` is set to `OIDC`. This will prevent logs in the kube-apiserver associated with not being able to connect to the oauth-apiserver, which we know should not be running when OIDC is enabled.
```

Comment:
the kube-apiserver should avoid using the current state of API resources to control its behavior.  Doing so means that manipulation via the API changes runtime behavior in a way that we cannot be confident when a cluster is or is not enforcing behavior.  Additionally, it means that code must dynamically react which is generally more complex to build.

Reply:
Instead, the admission chain can be configured by the operator

Reply:
Not controlling the behavior of the kube-apiserver with api resources makes sense to me. I don't think configuration of the admission chain itself will be sufficient to skip this post-start hook.

Would checking for all the criteria (apiserver flags, etc.) to know that the kube-apiserver is _not_ using the openshift oauth-apiserver for auth decisions be possible and sufficient?

Reply:
Updated with something that I thought would be sufficient based on information available to the kube-apiserver without relying on other api resources in [c32abf3](https://github.com/openshift/enhancements/pull/1726/commits/c32abf3d79d816cb71fcfd653c977974779e5a71)

Reply:
Another thought - add a new flag for enabling/disabling the post-start hook that is set by the cluster operators